stan/math/rev/core/var.hpp

```diff
@@ -102,24 +102,14 @@ class var_value {
   inline const auto& val() const { return vi_->val_; }
 
   /**
-   * Return a const reference of the derivative of the root expression with
+   * Return a reference of the derivative of the root expression with
    * respect to this expression.  This method only works
    * after one of the `grad()` methods has been
    * called.
    *
    * @return Adjoint for this variable.
    */
-  inline const auto& adj() const { return vi_->adj_; }
-
-  /**
-   * Return a reference to the derivative of the root expression with
-   * respect to this expression.  This method only works
-   * after one of the `grad()` methods has been
-   * called.
-   *
-   * @return Adjoint for this variable.
-   */
-  inline auto& adj() { return vi_->adj_; }
+  inline auto& adj() const { return vi_->adj_; }
```

Comment:
We never really respected const for the adjoint before.

I needed to make this change so something like this would work:
```cpp
myfunc(const var& a) {
  ...
  reverse_pass_callback([=]() {
    ...
    a.adj() += whatever;
  });
}
}
```

Reply:
@SteveBronder so yah know

Reply:
That feels weird to me. It's happening because you are taking in a `const &` so it calls the `const` version that gives back... a `const &`! imo the answer is to not use `const var&` when you need to modify the adjoint like this. Or maybe we should make `adj_` `mutable` in `vari_value? We do that for the read/write event stacks in `matrix_cl<>`.

Reply:
> is to not use const var& when you need to modify the adjoint like this

Yeah that's why I was giving it a shoutout cause it's going to look weird on review, but I think we practically make this assumption everywhere.

How this worked with the old stuff is that we just copied vari pointers out of everything. Those pointers themselves aren't constant so we were free to modify the stuff they contain.

Reply:
Why don't we just get rid of the `const` version of `adj()`?

Reply:
Or make `adj_` in the vari implementation `mutable` so even if we have a `const var` or whatever the `adj_` is still mutable